# Lab 2 - Hands-on with Transcriptomics Data

# Part 1: Understand The Dataset: Red and Blue Light Effect on Artemisinin Accumulation in _Artemisia annua_ L. 

## About Artemisinin

Artemisinin is a compound that was first identified and extracted from the plant, _Artemisia annua_ L.  (i.e., sweet wormwood). The compound is known for its anti-plasmodic properties, making it an excellent treatment for malaria. As of 2018, therapies oriented around artemisinin were the gold standard as far as malaria treatment is concerned (Dondorp et al. 2009). 

Also in 2018, there was growth in research demonstrating the compound having anti-tumor, anti-parasitic, anti-fibrosis, and anti-arrhythmic properties (Bin & Hong 2010). These properties have increased demand for the compound due to its potential for treating a wider variety of diseases and ailments (Peplow 2013). 

However, given that artemisinin can only be extracted directly from _Artemisia annua_ L, which already has low quantities of the compound in most varieties, there is a major bottleneck to scaling production of the compound. While there have been attempts to chemically synthesize, synthesize via cell culture, and heterologously synthesize via _Escherichia coli_ or yeast (Paddon et al. 2013) on artemisinin, they too pose barriers to being viable avenues for scaled production.

As such, the main focus has been to increase/improve artemisinin content in Artemisia annua. Research investigating the impact on abiotic factors, particularly light, has been promising with UV B and C radiation improving artemisinin content in A. annua despite these spectra harming overall growth and development (Rai et al. 2011). 

For that reason, the study from which this portfolio project originated from opted to explore other light spectra and found that red and blue light enhanced artemisinin content by upregulating expression of genes involved in artemisinin biosynthesis (e.g., ADS and CYP71AV1). 

**References:**

- Dondorp, A.M.; Nosten, F.; Yi, P.; Das, D.; Phyo, A.P.; Tarning, J.; Lwin, K.M.; Ariey, F.; Hanpithakpong, W.;
Lee, S.J.; et al. Artemisinin Resistance in Plasmodium falciparum Malaria. N. Engl. J. Med. 2009, 361, 455–467. [\[CrossRef\]](http://dx.doi.org/10.1056/NEJMoa0808859)

- Bin, L.; Hong, Z. Advances on the study of the pharmacological effects of artemisinin and its derivatives.
Chin. J. Clin. Pharmacol. Ther. 2010, 15, 572–576.

- Peplow, M. Malaria drug made in yeast causes market ferment. Nature 2013, 494, 160–161. [\[CrossRef\]](http://dx.doi.org/10.1038/494160a)

- Paddon, C.J.; Westfall, P.J.; Pitera, D.J.; Benjamin, K.; Fisher, K.; McPhee, D.; Leavell, M.D.; Tai, A.; Main, A.;
Eng, D.; et al. High-level semi-synthetic production of the potent antimalarial artemisinin. Nature 2013, 496,
528–532. [\[CrossRef\]](http://dx.doi.org/10.1038/nature12051)

- Rai, R.; Meena, R.P.; Smita, S.S.; Shukla, A.; Rai, K.; Pandey-Rai, S. UV-B and UV-C pre-treatments
induce physiological changes and artemisinin biosynthesis in Artemisia annua L.—An antimalarial plant.
J. Photochem. Photobiol. B Biol. 2011, 105, 216–225. [\[CrossRef\]](http://dx.doi.org/10.1016/j.jphotobiol.2011.09.004)



## Experimental Design

- **Organism:** _Artemisia annua_ L. 'Halqing 1' (wild-type)
- **Observation unit:** approximately 0.8 g of randomly selected aboveground parts
- **Conditions:**
    - LED red light (670 nm)
    - LED blue light (470 nm)
    - LED far-red light (735 nm)
    - white light

      
- **Replicates:** three biological replicates per treatment (10 seedlings per repeat)
- **Sequencing:** Illumina Xten sequencing system

  **Reference:** Zhang, D., Sun, W., Shi, Y., Wu, L., Zhang, T., & Xiang, L. Red and Blue Light Promote the Accumulation of Artemisinin in Artemisia annua L. Molecules, 23(6), 1329. [\[CrossRef\]](https://doi.org/10.3390/molecules23061329)


| Sample Name | SRA Accession | Condition | Replicate |
|-------------|---------------|-----------|-----------|
| B3 | SRR6808226 | LED blue light | 3 |
| B1 | SRR6808227 | LED blue light | 1 |
| B2 | SRR6808228 | LED blue light | 2 |
| RL2 | SRR6808229 | LED red light | 2 |
| RL3 | SRR6808230 | LED red light | 3 |
| WL1 | SRR6808231 | white light | 1 |
| WL2 | SRR6808232 | white light | 2 |
| WL3 | SRR6808239 | white light | 3 |
| RL1 | SRR6808240 | LED red light | 1 |

# Part 4: Exploring FASTQ Files

## View raw FASTQ data

In [6]:
cd ~/artemisia-annua-rna-seq-project/raw_data

]666;vte.shell.preexec!\


: 1

In [7]:
# View the first 12 lines (3 reads) of a FASTQ file
head -12 SRR6808226_1.fastq

# Use this if compressed:
# gzip -dc raw_data/KO_1_SRR10045016_1.fastq.gz | head -12
# gzip -dc works on both macOS and Linux

]666;vte.shell.preexec!\@SRR6808226.1 1 length=150
CTTTTTGCAAGCCGCTTCAAGTAAAGCATCTTTGGATGGTGAACTAGTATCTGAGTTGACATTATCTGCACTTTCCTGAAGAACCAGTGCCCAAGTATAAAGTGCATCATAATCATCTGGATTTCTCTCTATTGCACTTGCATACCTTTT
+SRR6808226.1 1 length=150
AAAFFJJJFFJJJ7FF7JFJ-AFJJ-<AFJFJF-<AFJJJJFFAAFAAJ7FF<J-<7F<F<-FJAJJ<<FJF-7FAAJJJAAAFJJJ-7AJJJJ--A-FJJJ<-AF-AJFJJ<AJ7AFJ7-77FJFF-<F--FAAJJ77AJFFJJJ--<F
@SRR6808226.2 2 length=150
GTGCAAGTGTGGATCAAGCTGCAAGTGTGATCCTTGCAACTGCTAAGCATGATAGGGAACTATAACCCGAACAACCTCTTGTGTTTTTTTATGTACTGTATTAAATAAAACAAAACCTATCTGTTATTCGTATTTGTAATGGGTTTCTAT
+SRR6808226.2 2 length=150
A-AAAAJJFAFJJJJJJ<F-FFJJJJJJJ-7FJJJFJJJJJA<-<JAFFFJJ7F-FFJJJ-FJJJJJJJAFJJJFFJJJJJJJF7<7F<7JJJJJJA-AF7JFFJFJJJJJJJJJJJAFFAJJ7JF---AAJFJAFFFJJFA---A<-A7
@SRR6808226.3 3 length=150
TGCACATATATAAGAGGTATGTGCTGAGTCATCCTGCTATAATGTCTCGTCACCCGATACACAGTTTCAGGTACATATTCTAACACAACATTGAGGTACACTGCGTTCTTTCGGTTGTTGAATAGAAACAATTTTTTAAGGGAACAATAT
+SRR6808226.3 3 length=150
AAFFFJJJJJFJJAJJJJF<FFJJF-FA<AJ7FFF7F<F-FF<AFF<FFFJJJJJJF

: 1

In [10]:
# Count total lines in a file (each read = 4 lines)
echo "Total lines in B3, paired end 1:"
wc -l SRR6808226_1.fastq

# Use this if compressed: 
# gzip -dc raw_data/KO_1_SRR10045016_1.fastq.gz | wc -l

]666;vte.shell.preexec!\Total lines in B3, paired end 1:
]666;vte.shell.preexec!\105833928 SRR6808226_1.fastq


: 1

In [9]:
# Count total number of reads in a file (each read = 4 lines)
echo "Total reads in B3, paired end 1:"
wc -l SRR6808226_1.fastq | awk '{print $1/4}'

# Use this if compressed:
# gzip -dc raw_data/KO_1_SRR10045016_1.fastq.gz | wc -l | awk '{print $1/4}'

]666;vte.shell.preexec!\Total reads in B3, paired end 1:
]666;vte.shell.preexec!\26458482


: 1

In [11]:
# View a random sample of reads
head -400 SRR6808226_1.fastq | tail -8

# Use this if compressed:
# gzip -dc raw_data/KO_1_SRR10045016_1.fastq.gz | head -400 | tail -8

]666;vte.shell.preexec!\@SRR6808226.99 99 length=150
CTAGTTCCACGTGGTGTGCCGGTAGGGCCATAGGTCTGTTTGTTGGTTATAAATGACAATGATGTAACGTATGTATTCCCAGCTTCATTAGTTGCGTAAGTTCCGCTAAACCCAGTAATGTGTTCATCATCTTCAAAATTAAACGTGTCA
+SRR6808226.99 99 length=150
AAAFFJJJJJJJJJJJJFJJFJFJJJJJJJJJJJFFJFFJ7F-FFJJFJFJJJF<JJJJ<FJFJAJJJFAJFFFJJJFJJJJJFFJJJJJJJJFJJJJJJFA-AJJJJJJJJJFFFJA7AFFFFJFFFJJJAAFFJJJAAJJJJ<F7<AF
@SRR6808226.100 100 length=150
CGAAGATGTCGACCGTTTACTCCCAATGCAAAAAGGGCTTGTTTCAACTAAGTTTTTATTCGGGCTTTTAAAAACCGCGTTGATCATCCGGGTCGACCCATCATGCATATTGAATTTAGAAAAAAGGATTGGGATGCAGTTAGATCAAGC
+SRR6808226.100 100 length=150
A<AFFJJJJJJJFJJFJFJJFJJJJJFJJJJJJJJJJJJ-FJJJJJJJ7FFJF77--FJFJ7JJJ<FJFJFJJJJJJJJJFJFJJFJJJJJJFJJJJJJFJJF-AJJAFFJJJJJJJJFJJJJFFAJF-7AJ-AJJJJJFJJ<A-AFJAA


: 1

## Understand paired-end reads

In [12]:
# Compare read IDs between R1 and R2 (should match)
echo "=== First read from R1 (forward) ==="
head -4 SRR6808226_1.fastq

# Use this if compressed:
# gzip -dc raw_data/KO_1_SRR10045016_1.fastq.gz | head -4

echo ""
echo "=== First read from R2 (reverse) ==="
head -4 SRR6808226_2.fastq

# Use this if compressed:
# gzip -dc raw_data/KO_1_SRR10045016_2.fastq.gz | head -4

# The read IDs should be identical (or differ only by /1 and /2)

]666;vte.shell.preexec!\=== First read from R1 (forward) ===
]666;vte.shell.preexec!\@SRR6808226.1 1 length=150
CTTTTTGCAAGCCGCTTCAAGTAAAGCATCTTTGGATGGTGAACTAGTATCTGAGTTGACATTATCTGCACTTTCCTGAAGAACCAGTGCCCAAGTATAAAGTGCATCATAATCATCTGGATTTCTCTCTATTGCACTTGCATACCTTTT
+SRR6808226.1 1 length=150
AAAFFJJJFFJJJ7FF7JFJ-AFJJ-<AFJFJF-<AFJJJJFFAAFAAJ7FF<J-<7F<F<-FJAJJ<<FJF-7FAAJJJAAAFJJJ-7AJJJJ--A-FJJJ<-AF-AJFJJ<AJ7AFJ7-77FJFF-<F--FAAJJ77AJFFJJJ--<F
]666;vte.shell.preexec!\
]666;vte.shell.preexec!\=== First read from R2 (reverse) ===
]666;vte.shell.preexec!\@SRR6808226.1 1 length=150
TGATGAGGAAGGTCGGTCAAGGCAACGGATTCTGACATTTGCAGCAAAAAGGTATGCAAGTGCAATAGAGAGAAATCCAGATGATTATGATGCACTTTATAATGGGGCACTGGTTCTTCAGGAAAGTGCAGAGAATGTCAACTCAGATAC
+SRR6808226.1 1 length=150
<<AAFJFJJJJ<AFJJJJFJ<FFJFJJJJJFF<JJFJ-<FJJF<FJFFFJJJFFAAJ-AJJJJJFAA7F7-<7FJAFJFJJ<JJ-AF-FJ<AJJJ<A7AAFF---77-7A)FFA-AA7F<AJFJJJJFJJJJ)AF-<FJFJJ<7<FJ<<A


: 1

# Part 5: Using SeqKit for FASTQ Exploration

In [13]:
# activate environment with seqkit
mamba activate genomics

]666;vte.shell.preexec!\


: 1

## Get basic statistics with SeqKit

In [14]:
# Get statistics for one sample
seqkit stats SRR6808226_1.fastq --threads 16

]666;vte.shell.preexec!\file                format  type    num_seqs        sum_len  min_len  avg_len  max_len
SRR6808226_1.fastq  FASTQ   DNA   26,458,482  3,968,772,300      150      150      150


: 1

In [15]:
# Get more detailed statistics for one sample
seqkit stats -a SRR6808226_1.fastq --threads 16

]666;vte.shell.preexec!\file                format  type    num_seqs        sum_len  min_len  avg_len  max_len   Q1   Q2   Q3  sum_gap  N50  N50_num  Q20(%)  Q30(%)  AvgQual  GC(%)  sum_n
SRR6808226_1.fastq  FASTQ   DNA   26,458,482  3,968,772,300      150      150      150  150  150  150        0  150        1   98.17   94.83    28.29  42.49      0


: 1

In [17]:
mkdir -p seqkit_output

]666;vte.shell.preexec!\


: 1

In [66]:
# Get more detailed stats for All samples
seqkit stats -a -T *.fastq --threads 16 > seqkit_output/seqkit_stats.tsv

]666;vte.shell.preexec!\processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed files:  0 / 18 [--------------------------------------] ETA: 0s
processed fi

: 1

**Note 4/3/2026:** _Not showing cell output because when I ran the first time Jupyter had to stop running it from crashing. Will have to run this line in terminal on my local computer._

In [67]:
head -20 seqkit_output/seqkit_stats.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len	Q1	Q2	Q3	sum_gap	N50	N50_num	Q20(%)	Q30(%)	AvgQual	GC(%)	sum_n
SRR6808226_1.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	98	95	28.29	42.49	0
SRR6808226_2.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	97	92	26.22	42.53	0
SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	99	96	29.25	42.49	0
SRR6808227_2.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	97	93	26.85	42.51	0
SRR6808228_1.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	99	96	29.23	42.51	0
SRR6808228_2.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	97	93	26.72	42.53	0
SRR6808229_1.fastq	FASTQ	DNA	24078723	3611808450	150	150.0	150	150	150	150	0	150	1	98	95	28.39	42.51	0
SRR6808229_2.fastq	FASTQ	DNA	24078723	3611808450	150	150.0	150	150	150	150	0	150	1	96	91	25.39	42.59	0
SRR6808230_1.fastq	FASTQ	DNA	

: 1

## Understand Stats on the fastq files

SeqKit reports multiple metrics on raw fastq files, open the `seqkit_stats.tsv` file and investigate the columns. 

# Part 6: Exploring Genome FASTA with SeqKit

SeqKit can also be used to explore and manipulate reference genome FASTA files. 

## Get genome statistics

In [20]:
cd ~/artemisia-annua-rna-seq-project/reference

]666;vte.shell.preexec!\


: 1

In [21]:
ls

]666;vte.shell.preexec!\GCA_003112345.1_ASM311234v1_genomic.fna
GCA_003112345.1_ASM311234v1_genomic.fna.seqkit.fai
GCA_003112345.1_ASM311234v1_genomic.gtf
GCA_003112345.1_ASM311234v1_rna_from_genomic.fna
seqkit_output
tx2gene.tsv


: 1

In [22]:
# Get detailed statistics on the reference genome file
seqkit stats GCA_003112345.1_ASM311234v1_genomic.fna

]666;vte.shell.preexec!\file                                     format  type  num_seqs        sum_len  min_len  avg_len  max_len
GCA_003112345.1_ASM311234v1_genomic.fna  FASTA   DNA     39,400  1,792,856,094    2,001   45,504  767,701


: 1

In [23]:
# Get detailed statistics including GC content
seqkit stats -a GCA_003112345.1_ASM311234v1_genomic.fna

]666;vte.shell.preexec!\file                                     format  type  num_seqs        sum_len  min_len  avg_len  max_len       Q1      Q2      Q3  sum_gap      N50  N50_num  Q20(%)  Q30(%)  AvgQual  GC(%)        sum_n
GCA_003112345.1_ASM311234v1_genomic.fna  FASTA   DNA     39,400  1,792,856,094    2,001   45,504  767,701  5,850.5  19,668  61,062        0  104,891    4,996       0       0        0  32.44  147,459,125


: 1

## Extract sequence regions

In [26]:
# Get the first 100 bases of the reference genome
seqkit subseq -r 1:100 GCA_003112345.1_ASM311234v1_genomic.fna > seqkit_output/first_100_bases_GCA_003112345.1_ASM311234v1_genomic.txt

]666;vte.shell.preexec!\[INFO] create or read FASTA index ...
[INFO] read FASTA index from GCA_003112345.1_ASM311234v1_genomic.fna.seqkit.fai
[INFO]   39400 records loaded from GCA_003112345.1_ASM311234v1_genomic.fna.seqkit.fai


: 1

**Note 4/3/2026:** _Not showing cell output because when I ran the first time Jupyter had to stop running it from crashing. Will have to send these lines to an output text file._

In [27]:
# Extract a specific region
seqkit subseq -r 3827:4962 GCA_003112345.1_ASM311234v1_genomic.fna > seqkit_output/region_3827_4962_GCA_003112345.1_ASM311234v1_genomic.txt

]666;vte.shell.preexec!\[INFO] create or read FASTA index ...
[INFO] read FASTA index from GCA_003112345.1_ASM311234v1_genomic.fna.seqkit.fai
[INFO]   39400 records loaded from GCA_003112345.1_ASM311234v1_genomic.fna.seqkit.fai


: 1

**Note 4/3/2026:** _Not showing cell output because when I ran the first time Jupyter had to stop running it from crashing. Will have to send these lines to an output text file._

In [28]:
# Get the length of each sequence in a FASTA file
seqkit fx2tab -l -n GCA_003112345.1_ASM311234v1_genomic.fna > seqkit_output/length_each_sequence_GCA_003112345.1_ASM311234v1_genomic.txt

]666;vte.shell.preexec!\


: 1

**Note 4/3/2026:** _Not showing cell output because when I ran the first time Jupyter had to stop running it from crashing. Will have to send these lines to an output text file._

## Calculate GC content and composition

In [29]:
# Calculate GC content for the genome
seqkit fx2tab -n -g GCA_003112345.1_ASM311234v1_genomic.fna > seqkit_output/GC_content_GCA_003112345.1_ASM311234v1_genomic.fna.txt

]666;vte.shell.preexec!\


: 1

**Note 4/3/2026:** _Not showing cell output because when I ran the first time Jupyter had to stop running it from crashing. Will have to send these lines to an output text file._

In [30]:
cd seqkit_output

]666;vte.shell.preexec!\


: 1

In [31]:
ls

]666;vte.shell.preexec!\GC_content_GCA_003112345.1_ASM311234v1_genomic.fna.txt
first_100_bases_GCA_003112345.1_ASM311234v1_genomic.txt
length_each_sequence_GCA_003112345.1_ASM311234v1_genomic.txt
region_3827_4962_GCA_003112345.1_ASM311234v1_genomic.txt


: 1

In [32]:
head -10 first_*.txt

]666;vte.shell.preexec!\>PKPP01000001.1 Artemisia annua cultivar Huhao1# QH_S00002, whole genome shotgun sequence
TTGTTGATATGGAGTATTTATCCCTGTGTCTAGATTTACGGTTTGAAGACTCAGGAAACT
CTCATTAAGCGATCAACGTAGCATGATCATCAAAAGCATG
>PKPP01000002.1 Artemisia annua cultivar Huhao1# QH_S00005, whole genome shotgun sequence
ttccgagtccacaagtgagactacattcgtgtagattaggatttactacatgataagagg
cattaggataacccatgttttacacctgtattataattac
>PKPP01000003.1 Artemisia annua cultivar Huhao1# QH_S00006, whole genome shotgun sequence
catttcccgagtcaacgactctttttatatccaaataacgttaataatttataagtatag
atattaatatgattggatatatcattcacgttgtccatta
>PKPP01000004.1 Artemisia annua cultivar Huhao1# QH_S00008, whole genome shotgun sequence


: 1

In [33]:
head -10 length_*.txt

]666;vte.shell.preexec!\PKPP01000001.1 Artemisia annua cultivar Huhao1# QH_S00002, whole genome shotgun sequence	691984
PKPP01000002.1 Artemisia annua cultivar Huhao1# QH_S00005, whole genome shotgun sequence	653913
PKPP01000003.1 Artemisia annua cultivar Huhao1# QH_S00006, whole genome shotgun sequence	629988
PKPP01000004.1 Artemisia annua cultivar Huhao1# QH_S00008, whole genome shotgun sequence	596958
PKPP01000005.1 Artemisia annua cultivar Huhao1# QH_S00011, whole genome shotgun sequence	580407
PKPP01000006.1 Artemisia annua cultivar Huhao1# QH_S00012, whole genome shotgun sequence	576155
PKPP01000007.1 Artemisia annua cultivar Huhao1# QH_S00010, whole genome shotgun sequence	585043
PKPP01000008.1 Artemisia annua cultivar Huhao1# QH_S00014, whole genome shotgun sequence	526215
PKPP01000009.1 Artemisia annua cultivar Huhao1# QH_S00007, whole genome shotgun sequence	601491
PKPP01000010.1 Artemisia annua cultivar Huhao1# QH_S00015, whole genome shotgun sequence	521217


: 1

In [34]:
head -20 region_*.txt

]666;vte.shell.preexec!\>PKPP01000001.1 Artemisia annua cultivar Huhao1# QH_S00002, whole genome shotgun sequence
TCACAAGGTTGCGCCTGACAAAACAATATTTCTCATACTCTATCCTTAACCATGTCTGAC
CTGGTATCCAATTCGGATCCTTATAAGTTcctcaaaatcaaacaaaacccaGACGGGTCA
TTAACCCGACTCGCCCCACTTCCATCCATACCCGCAACACCCCAACTCACTACCGAGTCA
CAACTGGTTCTCTCTAAAGATATACCACTCAACACCACCAATGCCACCTTTGTCCGTCTC
TTTCGGCCTATAATATCGAAAAGTACTAATGGAAAACTACCCATTATTATTTACTTCCAT
GGTGGGGGCTTTGTGTTCTTCAGCGCTACATCATTTCCTTTCCACGAGTCATGCACCGCG
GTGTCTGCTAAGACACCTGCGGTCGTGATCTCTGTTGAGTATCGTCTCGCACCCGAACAC
CGCCTGCCTGCTGCGTACGAGGACGCCATGCAGGCGGTGTTGTGGGTGCGTGATCAGGCT
TTAGGGGTTAATGGTTGTGATGCATGGTTGAGTGAGTTTGGTGATTTTAGTAAAGTGTTT
TTAATGGGCAGTAGTTCAGGAGGAAATATAGTTTATAATGCTGGTTTACGAGCCCTTGAT
CTTGATCTGAGTGTGATCAAGATCGCGGGGCTCATAATGAACCAGCCTTTTTTTGGTGGG
GTCAAGGACACGGAGTCTGAGGTTAGATTGGTAAATGACCGTATTGTCCCTTTAATGGCA
AGTCGTCTCATGTGGTCTCTTGCACTCCGAGAAGGGTGTAACAGAGACCACGAGTATTGT
AACCCGTTACGAGATCAAGATGAATCTCGTAATGAGAAAATTAAGCGTTTGCCGAAATGT
ATGGTTAGGGGTTACGCTGGGGACCCACTG

: 1

# Part 7: Useful SeqKit Commands

Here are additional SeqKit commands useful for bioinformatics workflows. 

## Subsample reads

In [35]:
cd ~/artemisia-annua-rna-seq-project/raw_data

]666;vte.shell.preexec!\


: 1

In [36]:
mkdir subsampled_reads_test

]666;vte.shell.preexec!\mkdir: subsampled_reads_test : Le fichier existe


: 1

In [37]:
ls

]666;vte.shell.preexec!\SRR6808226_1.fastq
SRR6808226_2.fastq
SRR6808227_1.fastq
SRR6808227_2.fastq
SRR6808228_1.fastq
SRR6808228_2.fastq
SRR6808229_1.fastq
SRR6808229_2.fastq
SRR6808230_1.fastq
SRR6808230_2.fastq
SRR6808231_1.fastq
SRR6808231_2.fastq
SRR6808232_1.fastq
SRR6808232_2.fastq
SRR6808239_1.fastq
SRR6808239_2.fastq
SRR6808240_1.fastq
SRR6808240_2.fastq
artemisia-annua-portfolio-project-sra-run-table-light-comparison.csv
artemisia-annua-portfolio-project-sra-run-table-tissue-comparison.csv
seqkit_output
subsampled_reads_test


: 1

In [38]:
# Randomly sample 50 reads from a FASTQ file
seqkit sample -n 50 SRR6808226_1.fastq -o subsampled_reads_test/subsampled_50_SRR6808226_1.fastq

]666;vte.shell.preexec!\[INFO] sample by number
[INFO] loading all sequences into memory...
[INFO] 44 sequences outputted


: 1

In [39]:
head -20 subsampled_reads_test/subsampled_50_*.fastq

]666;vte.shell.preexec!\@SRR6808226.543772 543772 length=150
GGTGAATCAGCTTCAAGAAGAGATAAGGGAACTCAAACAGAAACATAAGCAAGAGCTCCACGAGGCATTGACTCACCGTGAACTCCTGCAGCAGGATGTAGAACGTGAAAAAGCTGCACGTTTGGAATTAGAAAAAGCAGCTCACCTTCA
+
AAAAFJJFJFFJAFFFJJJFJFAFJA<AJF<AFFJJFFJFAJJJJFJJFFFJFFFJFJFJJFJJJJJJF<JJJFFAJJAAJJJAF7<<<F<AJJJ77AJJFFAJJJJJJJJFA<F-<AAA<<<FJFA<JJJJJJFJJAJFAAAJFJA-<A
@SRR6808226.1134852 1134852 length=150
CAACAGTTCCGTCGGGAAGTACAGATCCGAGAAGCGTCCCGATGACTCGCTCAGCTGAGTCGGGTCGTCGGACGAAACAATCGCAGATGTTGAAGATAACGAGTGGATGAACCTTGGCTGAGAGGCTGGTTGATAATGGGAAGAACTGCA
+
AAAFFJJFFFJJAFJJFJJFJFJFJAJFFJAJFFJJAJFJJJJJJJJFJFFFJFJ7JAF-FFJFFFJJFFFFFJJFAFFFJJJJJJ<JJFFFJFJAAFAAJFJJJ<F<FJFJFFJFF<FA7F-FJ<A7JAAFJ<-A-AJAAF<AAFFJF<
@SRR6808226.1558621 1558621 length=150
CACCATTCAAGAAATTTGCATCAAAGCGGGAGGAGTGGGCCCTTCAGAACAGATATATTAGTCCTGGCCCAATCCAGTTTACAGGACCAGTGGCTAATATCTCTAACCACACCTTATTATTGGAGCTTGGAGTTCAAGTCTAGTTACATC
+
AAFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ

: 1

In [40]:
# Sample a proportion (2%) of reads
seqkit sample -p 0.02 SRR6808226_1.fastq -o subsampled_reads_test/subsampled_2pct_SRR6808226_1.fastq

]666;vte.shell.preexec!\[INFO] sample by proportion
[INFO] 530623 sequences outputted


: 1

In [41]:
head -20 subsampled_reads_test/subsampled_2pct_*.fastq

]666;vte.shell.preexec!\@SRR6808226.50 50 length=150
GGTTTTCCCTCTTTTGCTTTGGTCAACAAAGCTACTGGACAAGCAATTAAGCACTCAAGTTCTGGTCAGCCGGTTCAGTTGGTCGAATTTAATCCAGATAAGCTTGATGAGTCCGTTCTTTGGACCCAGAGTAAGGACTTGGGAGACGGT
+
AAAFFJJJJJJJJJJJJFFJJJFJFJJJJJJJJJJAA<JFFJFJJJJJJJJJJJFJJJJJJJJJFJJJAJJJJAJFJJJJJJJJJJJJJJJJJJJJJJJJJFFJJJJJFJJJJJFFFJFFJFJFJFFJJJJFFJJJJJJJJJJJF<FAJF
@SRR6808226.52 52 length=150
CCGGAATCCAGAGCCGCCATCTTGGCCCAAGATTTGACACCTAACGCCGACGGATCATTGAAGAACAATTTCCAAACCGAAAATGGAATCAGACAAGAAGAAGTCAGATACTTGAAGGCTGGCCCAGAAGGACCAGTATCGGTGGTCCAG
+
AAFFFJJJJJJJFJJJJJJAJJJJFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFAFJJJJJJJJJJ7AJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJFJJJJAJJJJFJJJFJJJJJJJJJJJJFJJJJJAJ-FJFFJJFFJJJ
@SRR6808226.144 144 length=150
CAAATATATCTAAAGAGACCAAACAATTCACCTACAAGGATGGCTCCCAGTTTGTTTTCATGGACCTGACTTCGTACGAGGAAGTGCGTCTTAACGAGTCAGATATGGGAGACAAGACTAAATGGCTAAAAGAAGGCATGGATTGCATTT
+
AAAFAJFJ<FFJJJJJJJJJJJJJAFFAJFJJAJJJJJJJ<JAA<JJJFJJJJJF---<J-7FF<AJJFJ-<A-77F<AJJJFFJJFFFJ-AAJJAJJAJFFJF7AFJAFJFJFF7-A-7AJAA

: 1

In [42]:
# Sample with a specific seed for reproducibility
seqkit sample -n 50 -s 42 SRR6808226_1.fastq -o subsampled_reads_test/subsampled_seed42_SRR6808226_1.fastq

]666;vte.shell.preexec!\[INFO] sample by number
[INFO] loading all sequences into memory...
[INFO] 50 sequences outputted


: 1

In [43]:
head -20 subsampled_reads_test/subsampled_seed42_*.fastq

]666;vte.shell.preexec!\@SRR6808226.2023706 2023706 length=150
CTCAATTGGAATGCGTGAACTTCCCATCCCTGAACACATGGATATATTCCATCTCACTAGAGAAATTGAGGCTTCTGACATGTCCTCACTTGAGGCTGTCATGAATTGTGACGAAGAGAGGCTGAAATTAGAAGCTGAAGCTGAACGTTT
+
AAFFFJJJJJJFJJJJJJFFJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJ<FFJJAJFFJJJJJJJJJJJFJAF7AJJJJFJJJJJJJJJJJJJ<FJJJAFF
@SRR6808226.2220436 2220436 length=150
TTAAAAACAAATATAACAATGGTCAAAAACAAAGCAGATTTTCTGACATATGTGTTCTTGAGGAAATAGATGAAGACATTTTCAATTTATCTAATGATTATTCTCTTGCTCATTGTGTAGCTGAAGACTTAAGAATGGGTGCAGGCATAG
+
AAAFFJJJJJJJJJJJJJAJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJFJJ<JJJJJFJJJJJJJJJJFJJJFFFFJJJJJJJAJJFJJFFJFFJFJJJ
@SRR6808226.2426638 2426638 length=150
CTCGAATCATTACCCGAAATTCAAGACCGGTACTTTAACTTACCAAGCGTCAATTCCATAAAAACCTTGATAAATGAAGACCAAAAAGTAAATATCCAAAAATATTATTCGGGTAATTATGATTGGGCTAGCATAGCCGCATTCGGGCAA
+
<<AAAFJJAJJJFJJJFJ7AFJJ<-JAFA<7F7-FJ<<FFJJF-<FJ-7-7FJ7JJJ--FA<JJJF--77FF-JFFFFAFJAJFJAJJ<FJF-F7-

: 1

**When to Subsample?**

Subsampling is useful for: 
- Testing pipelines quickly before running on full data
- Debugging issues with smaller datasets
- Creating balanced datasets when samples have unequal depths

## Search for sequence patterns

In [44]:
# Search for reads containing a specific sequence motif
seqkit grep -s -p "AGATCGGAAG" SRR6808226_1.fastq | head -20

]666;vte.shell.preexec!\@SRR6808226.4647 4647 length=150
AAGTGCTCTTGGTTGTTTGACATGGCCAAAACCATCCCATTCAGAATTCCAACAGTTGGCTAGAGAAAGCGAGTACGCAGCTTGGACCCTTGTTAATGGATACACCTTGAATCTTGAGATCGGAAGAGCATACGTCTGTACTTGAGTCAC
+
<AF-FFAJFAJFAJJ--7FJFJFJ<7JAJ<AAFFJJJFJ-<<FF7JFJ--7FA-7--<FJ7JF7<-AJJ---7-7-7A-7--7AF<A-7AAF7-FAF<F---<-7AA--FJ<F-77FFA<AAF7FA<<-7--77<JFJ-7----7-7FA7
@SRR6808226.7433 7433 length=150
AGTCAGTCGATTAAAAAAAAAAAAACAATAATAATAATGGCGGCAGCTTCTGCAGCAGCCGGATCACTCACCAGCAGCAGCAACAAACTCCCATGCCGATTCTTCCGATCTCCACGTGGTCTCGGCTTCGCTGACTCAATTGCTTCTCAC
+
AAFFFJAFAJJJJJJJJJJJJJJJJJJAJJJJJJJJJJFFFJFJJJJJJJFJJJ<FJJJJJAFJJFJFJJJJJJJFAFAAJAJJJJJJ<FJJAJFFJFFJJJ-7AJAFAF<JFAJF<-AFJ-F77A7)A<FA-<A7<JJ--<A--7-7-A
@SRR6808226.13066 13066 length=150
CCCAGGTTCCGTTGACGGCGACGGCAATGGCCGCGGTTTCGACGATTGGCGAGATCGGAAGAGCACACGTGTGAACTCAAGGTATCCTCCGGTATCTAGTATGAGGTATTGTGATAGTAAAAAAACCATAGTTAATACACAAAAATGTTC
+
AAAF7A7FJ<FAJAJAJFJFFJJAJFA<-AFAJJJJAAFFFFJF7<FA7FFA<-7--7-<<F--7<AFA---7F77-7-777-A-7---7---77-7----7---77-A7--

: 1

In [45]:
# Count reads containing adapter sequence (Illumina universal adapter)
seqkit grep -s -p "AGATCGGAAGAGC" SRR6808226_1.fastq | seqkit stats

]666;vte.shell.preexec!\file  format  type  num_seqs  sum_len  min_len  avg_len  max_len
-     FASTQ   DNA      1,598  239,700      150      150      150


: 1

## Sequence manipulation

In [51]:
# Get reverse complement of sequences
seqkit seq -r -p SRR6808226_1.fastq | head -8

]666;vte.shell.preexec!\[WARN] flag -t (--seq-type) (DNA/RNA) is recommended for computing complement sequences
@SRR6808226.1 1 length=150
AAAAGGTATGCAAGTGCAATAGAGAGAAATCCAGATGATTATGATGCACTTTATACTTGGGCACTGGTTCTTCAGGAAAGTGCAGATAATGTCAACTCAGATACTAGTTCACCATCCAAAGATGCTTTACTTGAAGCGGCTTGCAAAAAG
+
F<--JJJFFJA77JJAAF--F<-FFJF77-7JFA7JA<JJFJA-FA-<JJJF-A--JJJJA7-JJJFAAAJJJAAF7-FJF<<JJAJF-<F<F7<-J<FF7JAAFAAFFJJJJFA<-FJFJFA<-JJFA-JFJ7FF7JJJFFJJJFFAAA
@SRR6808226.2 2 length=150
ATAGAAACCCATTACAAATACGAATAACAGATAGGTTTTGTTTTATTTAATACAGTACATAAAAAAACACAAGAGGTTGTTCGGGTTATAGTTCCCTATCATGCTTAGCAGTTGCAAGGATCACACTTGCAGCTTGATCCACACTTGCAC
+
7A-<A---AFJJFFFAJFJAA---FJ7JJAFFAJJJJJJJJJJJFJFFJ7FA-AJJJJJJ7<F7<7FJJJJJJJFFJJJFAJJJJJJJF-JJJFF-F7JJFFFAJ<-<AJJJJJFJJJF7-JJJJJJJFF-F<JJJJJJFAFJJAAAA-A


: 1

In [52]:
# Extract only sequenced IDs
seqkit seq -n SRR6808226_1.fastq | head -10

]666;vte.shell.preexec!\SRR6808226.1 1 length=150
SRR6808226.2 2 length=150
SRR6808226.3 3 length=150
SRR6808226.4 4 length=150
SRR6808226.5 5 length=150
SRR6808226.6 6 length=150
SRR6808226.7 7 length=150
SRR6808226.8 8 length=150
SRR6808226.9 9 length=150
SRR6808226.10 10 length=150


: 1

In [53]:
# Get sequences within a length range
seqkit seq -m 60 -M 80 SRR6808226_1.fastq | seqkit stats

]666;vte.shell.preexec!\[WARN] you may switch on flag -g/--remove-gaps to remove spaces
file  format  type  num_seqs  sum_len  min_len  avg_len  max_len
-                          0        0        0        0        0


: 1

## Compare and deduplicate sequences

In [37]:
# Remove duplicate sequences (by sequence content)
seqkit rmdup -s SRR6808226_1.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] 12257516 duplicated records removed
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   14,200,966  2,130,144,900      150      150      150


: 1

**TODO (4/3/2026):** Will have to deduplicate all of the fastq files within my portfolio `raw_data` directory in order to keep everything consistent when conducting further downstream analyses.  

**UPDATE (4/12/2026):** Deduplicated the rest of the fastq files. I then re-ran all of the other code blocks in this jupyter notebook besides the blocks under the subsection titled ***Compare and deduplicate sequences*** for consistency's sake. 

In [3]:
# Remove duplicate sequences (by sequence content) for remaining samples
seqkit rmdup -s SRR680822[7-9]*.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] 100937344 duplicated records removed
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   66,556,196  9,983,429,400      150      150      150


: 1

In [4]:
# Remove duplicate sequences (by sequence content) for remaining samples
seqkit rmdup -s SRR680823[0,1,2,9]*.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] 107831636 duplicated records removed
file  format  type    num_seqs         sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   74,884,126  11,232,618,900      150      150      150


: 1

In [5]:
# Remove duplicate sequences (by sequence content) for remaining samples
seqkit rmdup -s SRR6808240*.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] 20243349 duplicated records removed
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   24,216,409  3,632,461,350      150      150      150


: 1

In [55]:
# Find common sequences between two files (useful for checking paired-end sync)
seqkit common -n SRR6808226_1.fastq SRR6808226_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808226_1.fastq
[INFO] read file 2/2: SRR6808226_2.fastq
[INFO] find common seqs ...
[INFO] 26458482 unique sequence headers found in 2 files, which belong to 26458482 records in the first file: SRR6808226_1.fastq
[INFO] 26458482 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   26,458,482  3,968,772,300      150      150      150


: 1

# Exercises

## Exercise 1: Read Analysis

1. What is the average read length across all samples? 

In [74]:
# Get an idea of the structure of the seqkit stats tsv file across all samples
head -40 seqkit_output/seqkit_stats.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len	Q1	Q2	Q3	sum_gap	N50	N50_num	Q20(%)	Q30(%)	AvgQual	GC(%)	sum_n
SRR6808226_1.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	98	95	28.29	42.49	0
SRR6808226_2.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	97	92	26.22	42.53	0
SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	99	96	29.25	42.49	0
SRR6808227_2.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	97	93	26.85	42.51	0
SRR6808228_1.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	99	96	29.23	42.51	0
SRR6808228_2.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	97	93	26.72	42.53	0
SRR6808229_1.fastq	FASTQ	DNA	24078723	3611808450	150	150.0	150	150	150	150	0	150	1	98	95	28.39	42.51	0
SRR6808229_2.fastq	FASTQ	DNA	24078723	3611808450	150	150.0	150	150	150	150	0	150	1	96	91	25.39	42.59	0
SRR6808230_1.fastq	FASTQ	DNA	

: 1

In [69]:
awk -F'\t' -v col=7 '
BEGIN{OFS="\t"} 
NR>1 {
sum += $col
count++
} END {
avg = sum / count
print "Average read length across all samples: " avg
}' seqkit_output/seqkit_stats.tsv

]666;vte.shell.preexec!\Average read length across all samples: 150


: 1

> The average read length across all samples is 150 bp. 

2. What sample has the highest number of reads? 

In [81]:
sort -k4,4nr seqkit_output/seqkit_stats.tsv | awk '
BEGIN{OFS="\t"} 
NR>= 1 && NR <= 2 {
print $0, NR
}'

# NR==1 prints only the first line, here I did NR>=1 && NR <= 2
# to print the paired reads of the same sample. 

]666;vte.shell.preexec!\SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	99	96	29.25	42.49	0	1
SRR6808227_2.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	97	93	26.85	42.51	0	2


: 1

> The sample with the highest number of reads was sample SRR6808227, or the first replicate of the blue light treatment group, B1. It had a total of 30,751,077 reads. 

3. Calculate the total sequencing depth (total bases) for blue light vs white light groups. 

In [82]:
cd ~/artemisia-annua-rna-seq-project/raw_data

]666;vte.shell.preexec!\


: 1

In [3]:
ls

]666;vte.shell.preexec!\SRR6808226_1.fastq
SRR6808226_2.fastq
SRR6808227_1.fastq
SRR6808227_2.fastq
SRR6808228_1.fastq
SRR6808228_2.fastq
SRR6808229_1.fastq
SRR6808229_2.fastq
SRR6808230_1.fastq
SRR6808230_2.fastq
SRR6808231_1.fastq
SRR6808231_2.fastq
SRR6808232_1.fastq
SRR6808232_2.fastq
SRR6808239_1.fastq
SRR6808239_2.fastq
SRR6808240_1.fastq
SRR6808240_2.fastq
artemisia-annua-portfolio-project-sra-run-table-light-comparison.csv
artemisia-annua-portfolio-project-sra-run-table-tissue-comparison.csv
seqkit_output
subsampled_reads_test


: 1

In [83]:
# Count total bases for blue light samples (SRR6808227, SRR6808228, SRR6808226) using awk only
BlueSamples=("SRR6808226" "SRR6808227" "SRR6808228")

function LoopTotalBases
{
m=${#BlueSamples[@]}
# comment out m if you decide to not use indexing method
total_bases=0

for (( i=0; i<m; i++ ));
# can also do for file in "${BlueSamples[@]}";
do
   # define paired-end reads
   R1="${BlueSamples[$i]}_1.fastq"
   R2="${BlueSamples[$i]}_2.fastq"
   
   # matching logic wherein bases are only added after checking the number of bases in read 1 matches that of read 2
   # count number of bases in read 1
   bases_r1=$(awk 'NR%4==2 {sum+=length($0)} END {print sum}' $R1)
   
   # count number of bases in read 2
   bases_r2=$(awk 'NR%4==2 {sum+=length($0)} END {print sum}' $R2)
   
   # compare number of bases of read 1 to read 2
   # additional logic for cases where the paired end read counts don't match
   # if the paired end read counts dont match (the difference is not zero), take the read count of the read with lower number of bases
   # better to take read with lower number of bases
   if [ "$bases_r1" -eq "$bases_r2" ]; then
     # add current number of bases to total
     # can also do total_bases=$((total_bases + bases)
     ((total_bases+=bases_r1))
   elif [ "$bases_r1" -gt "$bases_r2" ]; then
       ((total_bases+=bases_r2))
   elif [ "$bases_r1" -lt "$bases_r2" ]; then
       ((total_bases+=bases_r1))
   else
       :
   fi
done

echo "Total bases for blue light samples: $total_bases"
}

LoopTotalBases ${BlueSamples[@]}



]666;vte.shell.preexec!\
]666;vte.shell.preexec!\
]666;vte.shell.preexec!\Total bases for blue light samples: 12918979350


: 1

In [84]:
ls

]666;vte.shell.preexec!\SRR6808226_1.fastq
SRR6808226_2.fastq
SRR6808227_1.fastq
SRR6808227_2.fastq
SRR6808228_1.fastq
SRR6808228_2.fastq
SRR6808229_1.fastq
SRR6808229_2.fastq
SRR6808230_1.fastq
SRR6808230_2.fastq
SRR6808231_1.fastq
SRR6808231_2.fastq
SRR6808232_1.fastq
SRR6808232_2.fastq
SRR6808239_1.fastq
SRR6808239_2.fastq
SRR6808240_1.fastq
SRR6808240_2.fastq
artemisia-annua-portfolio-project-sra-run-table-light-comparison.csv
artemisia-annua-portfolio-project-sra-run-table-tissue-comparison.csv
seqkit_output
subsampled_reads_test


: 1

In [92]:
# Count total bases for blue light samples (SRR6808227, SRR6808228, SRR6808226) using seqkit and awk

# activate genomics mamba environment
mamba activate genomics

]666;vte.shell.preexec!\


: 1

In [93]:
# Check paired-end reads for B1
seqkit common -n SRR6808227_1.fastq SRR6808227_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808227_1.fastq
[INFO] read file 2/2: SRR6808227_2.fastq
[INFO] find common seqs ...
[INFO] 30751077 unique sequence headers found in 2 files, which belong to 30751077 records in the first file: SRR6808227_1.fastq
[INFO] 30751077 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   30,751,077  4,612,661,550      150      150      150


: 1

In [94]:
# Check paired-end reads for B2
seqkit common -n SRR6808228_1.fastq SRR6808228_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808228_1.fastq
[INFO] read file 2/2: SRR6808228_2.fastq
[INFO] find common seqs ...
[INFO] 28916970 unique sequence headers found in 2 files, which belong to 28916970 records in the first file: SRR6808228_1.fastq
[INFO] 28916970 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   28,916,970  4,337,545,500      150      150      150


: 1

In [95]:
# Check paired-end reads for B3
seqkit common -n SRR6808226_1.fastq SRR6808226_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808226_1.fastq
[INFO] read file 2/2: SRR6808226_2.fastq
[INFO] find common seqs ...
[INFO] 26458482 unique sequence headers found in 2 files, which belong to 26458482 records in the first file: SRR6808226_1.fastq
[INFO] 26458482 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   26,458,482  3,968,772,300      150      150      150


: 1

_All three samples in the blue light group have matching paired-end reads, so we can proceed with calculating the total depth for all of the samples based on the value of just one paired read for each sample._

In [96]:
# run seqkit for all blue light samples
seqkit stats SRR680822[6-8]_1.fastq --threads 16 -T > seqkit_output/blue_light_samples_total_depth.tsv

]666;vte.shell.preexec!\processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [

: 1

In [97]:
# Show output of total depth tsv file for blue light samples
head -10 seqkit_output/blue_light_samples_total_depth.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len
SRR6808226_1.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150
SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150
SRR6808228_1.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150


: 1

In [98]:
# Use awk to process through the sum_len column for each sample to calculate the total depth for the blue light treatment group
awk -F'\t' '{sum+=$5} END{print "Total number of bases for blue light samples: "sum}' seqkit_output/blue_light_samples_total_depth.tsv

]666;vte.shell.preexec!\Total number of bases for blue light samples: 12918979350


: 1

In [99]:
# Count total bases for white light samples (SRR6808231, SRR6808232, SRR6808239) using awk only
WhiteSamples=(SRR6808231 SRR6808232 SRR6808239)

function LoopTotalBases
{
m=${#WhiteSamples[@]}
# comment out m if you decide to not use indexing method
total_bases=0

for (( i=0; i<m; i++ ));
# can also do for file in "${BlueSamples[@]}";
do
   # define paired-end reads
   R1="${WhiteSamples[$i]}_1.fastq"
   R2="${WhiteSamples[$i]}_2.fastq"
   
   # matching logic wherein bases are only added after checking the number of bases in read 1 matches that of read 2
   # count number of bases in read 1
   bases_r1=$(awk 'NR%4==2 {sum+=length($0)} END {print sum}' $R1)
   
   # count number of bases in read 2
   bases_r2=$(awk 'NR%4==2 {sum+=length($0)} END {print sum}' $R2)
   
   # compare number of bases of read 1 to read 2
   # additional logic for cases where the paired end read counts don't match
   # if the paired end read counts dont match (the difference is not zero), take the read count of the read with lower number of bases
   # better to take read with lower number of bases
   if [ "$bases_r1" -eq "$bases_r2" ]; then
     # add current number of bases to total
     # can also do total_bases=$((total_bases + bases)
     ((total_bases+=bases_r1))
   elif [ "$bases_r1" -gt "$bases_r2" ]; then
       ((total_bases+=bases_r2))
   elif [ "$bases_r1" -lt "$bases_r2" ]; then
       ((total_bases+=bases_r1))
   else
       :
   fi
done

echo "Total bases for blue light samples: $total_bases"
}

LoopTotalBases ${WhiteSamples[@]}



]666;vte.shell.preexec!\
]666;vte.shell.preexec!\
]666;vte.shell.preexec!\Total bases for blue light samples: 10156788150


: 1

In [100]:
# Check paired-end reads for WL1
seqkit common -n SRR6808231_1.fastq SRR6808231_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808231_1.fastq
[INFO] read file 2/2: SRR6808231_2.fastq
[INFO] find common seqs ...
[INFO] 23119663 unique sequence headers found in 2 files, which belong to 23119663 records in the first file: SRR6808231_1.fastq
[INFO] 23119663 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   23,119,663  3,467,949,450      150      150      150


: 1

In [101]:
# Check paired-end reads for WL2
seqkit common -n SRR6808232_1.fastq SRR6808232_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808232_1.fastq
[INFO] read file 2/2: SRR6808232_2.fastq
[INFO] find common seqs ...
[INFO] 19885886 unique sequence headers found in 2 files, which belong to 19885886 records in the first file: SRR6808232_1.fastq
[INFO] 19885886 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   19,885,886  2,982,882,900      150      150      150


: 1

In [102]:
# Check paired-end reads for WL3
seqkit common -n SRR6808239_1.fastq SRR6808239_2.fastq | seqkit stats

]666;vte.shell.preexec!\[INFO] read file 1/2: SRR6808239_1.fastq
[INFO] read file 2/2: SRR6808239_2.fastq
[INFO] find common seqs ...
[INFO] 24706372 unique sequence headers found in 2 files, which belong to 24706372 records in the first file: SRR6808239_1.fastq
[INFO] 24706372 common/shared sequences saved to: -
file  format  type    num_seqs        sum_len  min_len  avg_len  max_len
-     FASTQ   DNA   24,706,372  3,705,955,800      150      150      150


: 1

_All three samples in the white light group have matching paired-end reads, so we can proceed with calculating the total depth for all of the samples based on the value of just one paired read for each sample._

In [103]:
# Count total bases for white light samples (SRR6808231, SRR6808232, SRR6808239) using seqkit and awk

# run seqkit for all white light samples
seqkit stats SRR680823[1,2,9]_1.fastq --threads 16 -T > seqkit_output/white_light_samples_total_depth.tsv

]666;vte.shell.preexec!\processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [--------------------------------------] ETA: 0s
processed files:  0 / 3 [

: 1

In [104]:
# Show output of total depth tsv file for white light samples
head -10 seqkit_output/white_light_samples_total_depth.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len
SRR6808231_1.fastq	FASTQ	DNA	23119663	3467949450	150	150.0	150
SRR6808232_1.fastq	FASTQ	DNA	19885886	2982882900	150	150.0	150
SRR6808239_1.fastq	FASTQ	DNA	24706372	3705955800	150	150.0	150


: 1

In [105]:
# Use awk to process through the sum_len column for each sample to calculate the total depth for the blue light treatment group
awk -F'\t' '{sum+=$5} END{print "Total number of bases for white light samples: "sum}' seqkit_output/white_light_samples_total_depth.tsv

]666;vte.shell.preexec!\Total number of bases for white light samples: 10156788150


: 1

> Total sequencing depth (total bases) for blue light group: 12918979350
> Total sequencing depth (total bases) for white light group: 10156788150

1. Find reads that start with 'N' (unknown base). How many are there in B1_1 and WL1_1? 

In [106]:
# Find reads that start with 'N' in B1_1
awk 'NR%4==2 && /^N/' SRR6808227_1.fastq | wc -l

]666;vte.shell.preexec!\0


: 1

In [107]:
# Find reads that start with 'N' in WL1_1
awk 'NR%4==2 && /^N/' SRR6808231_1.fastq | wc -l

]666;vte.shell.preexec!\0


: 1

> Total number of reads that start with 'N' in BL_1 sample: 0

> Total number of reads that start with 'N' in WL_1 sample: 0

2. What percentage of reads have a quality score below 20? 

In [108]:
# Find percentage of bases with quality scores below 20
awk 'BEGIN{OFS="\t"} $17 < 20 {print $0}' seqkit_output/seqkit_stats.tsv

]666;vte.shell.preexec!\


: 1

In [112]:
# Verification step: 
sort -k17,17nr seqkit_output/seqkit_stats.tsv | awk '
BEGIN{OFS="\t"} {
print $0, NR
}'

]666;vte.shell.preexec!\SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	99	96	29.25	42.49	0	1
SRR6808228_1.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	99	96	29.23	42.51	0	2
SRR6808226_1.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	98	95	28.29	42.49	0	3
SRR6808229_1.fastq	FASTQ	DNA	24078723	3611808450	150	150.0	150	150	150	150	0	150	1	98	95	28.39	42.51	0	4
SRR6808230_1.fastq	FASTQ	DNA	23645960	3546894000	150	150.0	150	150	150	150	0	150	1	98	95	28.42	42.58	0	5
SRR6808231_1.fastq	FASTQ	DNA	23119663	3467949450	150	150.0	150	150	150	150	0	150	1	98	95	28.54	42.65	0	6
SRR6808232_1.fastq	FASTQ	DNA	19885886	2982882900	150	150.0	150	150	150	150	0	150	1	98	95	28.51	42.73	0	7
SRR6808239_1.fastq	FASTQ	DNA	24706372	3705955800	150	150.0	150	150	150	150	0	150	1	98	95	28.53	42.80	0	8
SRR6808240_1.fastq	FASTQ	DNA	22229879	3334481850	150	150.0	150	150	150	150	0	150	1	98	95	28.26	42.56	0	9
SRR6808226_2.fastq	FASTQ	DNA	

: 1

> Lowest AvgQual is 24.49 while highest is 29.25, so the first awk filter was correct in that none of the reads have a quality score below 20. 

> Therefore, the percentage of reads that have a quality score below 20 is 0%.

3. Compare the GC content between blue light and white light samples. Is there a difference? 

In [113]:
# detailed seqkit output of blue light samples
seqkit stats -a SRR680822[6-8]_[1,2].fastq --threads 16 -T > seqkit_output/blue_light_samples_detailed.tsv

]666;vte.shell.preexec!\processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [

: 1

In [114]:
# detailed output of seqkit stats of blue light samples
head seqkit_output/blue_light_samples_detailed.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len	Q1	Q2	Q3	sum_gap	N50	N50_num	Q20(%)	Q30(%)	AvgQual	GC(%)	sum_n
SRR6808226_1.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	98	95	28.29	42.49	0
SRR6808226_2.fastq	FASTQ	DNA	26458482	3968772300	150	150.0	150	150	150	150	0	150	1	97	92	26.22	42.53	0
SRR6808227_1.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	99	96	29.25	42.49	0
SRR6808227_2.fastq	FASTQ	DNA	30751077	4612661550	150	150.0	150	150	150	150	0	150	1	97	93	26.85	42.51	0
SRR6808228_1.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	99	96	29.23	42.51	0
SRR6808228_2.fastq	FASTQ	DNA	28916970	4337545500	150	150.0	150	150	150	150	0	150	1	97	93	26.72	42.53	0


: 1

In [115]:
# Find average GC content across blue light samples
awk -F '\t' -v col=18 -v n=6 '
BEGIN{OFS="\t"}
NR>1 {
  sum += $col
} END {
avg = sum / n
print "Average GC content across blue light samples: " avg "%"
}' seqkit_output/blue_light_samples_detailed.tsv

]666;vte.shell.preexec!\Average GC content across blue light samples: 42.51%


: 1

In [116]:
# detailed seqkit output of white light samples
seqkit stats -a SRR680823[1,2,9]_[1,2].fastq --threads 16 -T > seqkit_output/white_light_samples_detailed.tsv

]666;vte.shell.preexec!\processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [--------------------------------------] ETA: 0s
processed files:  0 / 6 [

: 1

In [117]:
# detailed output of seqkit stats of white light samples
head seqkit_output/white_light_samples_detailed.tsv

]666;vte.shell.preexec!\file	format	type	num_seqs	sum_len	min_len	avg_len	max_len	Q1	Q2	Q3	sum_gap	N50	N50_num	Q20(%)	Q30(%)	AvgQual	GC(%)	sum_n
SRR6808231_1.fastq	FASTQ	DNA	23119663	3467949450	150	150.0	150	150	150	150	0	150	1	98	95	28.54	42.65	0
SRR6808231_2.fastq	FASTQ	DNA	23119663	3467949450	150	150.0	150	150	150	150	0	150	1	97	92	26.05	42.70	0
SRR6808232_1.fastq	FASTQ	DNA	19885886	2982882900	150	150.0	150	150	150	150	0	150	1	98	95	28.51	42.73	0
SRR6808232_2.fastq	FASTQ	DNA	19885886	2982882900	150	150.0	150	150	150	150	0	150	1	97	92	26.26	42.77	0
SRR6808239_1.fastq	FASTQ	DNA	24706372	3705955800	150	150.0	150	150	150	150	0	150	1	98	95	28.53	42.80	0
SRR6808239_2.fastq	FASTQ	DNA	24706372	3705955800	150	150.0	150	150	150	150	0	150	1	97	93	26.37	42.84	0


: 1

In [118]:
# Find average GC content across white light samples
awk -F '\t' -v col=18 -v n=6 '
BEGIN{OFS="\t"}
NR>1 {
  sum += $col
} END {
avg = sum / n
print "Average GC content across white light samples: " avg "%"
}' seqkit_output/white_light_samples_detailed.tsv

]666;vte.shell.preexec!\Average GC content across white light samples: 42.7483%


: 1

> There is a difference in GC(%) content between blue light and white light samples, but it is very marginal. On average, GC(%) content in blue light samples is 42.51 while in white samples it is 42.7483. Therefore, on average, white light samples have ever-so-slightly higher GC content than blue light samples with a difference of 0.2383%. 

## Exercise 3: SeqKit Exploration

Use `seqkit --help` to discover more commands. Try: 

1. `seqkit --help` -Revise the parameters

In [119]:
seqkit --help

]666;vte.shell.preexec!\SeqKit -- a cross-platform and ultrafast toolkit for FASTA/Q file manipulation

Version: 2.13.0

Author: Wei Shen <shenwei356@gmail.com>

Documents  : http://bioinf.shenwei.me/seqkit
Source code: https://github.com/shenwei356/seqkit
Please cite: https://doi.org/10.1002/imt2.191


Seqkit utilizes the pgzip (https://github.com/klauspost/pgzip) package to
read and write gzip file, and the outputted gzip file would be slighty
larger than files generated by GNU gzip.

Seqkit writes gzip files very fast, much faster than the multi-threaded pigz,
therefore there's no need to pipe the result to gzip/pigz.

Seqkit also supports reading and writing xz (.xz) and zstd (.zst) formats since v2.2.0.
Bzip2 format is supported since v2.4.0.

Compression level:
  format   range   default  comment
  gzip     1-9     5        https://github.com/klauspost/pgzip sets 5 as the default value.
  xz       NA      NA       https://github.com/ulikunitz/xz does not support.
  zstd     1-4

: 1

2. `seqkit grep -p "PATTERN"` -Search for specific sequences

In [121]:
# Extract sequences containing AGGCG (only in the positive strand)
cat SRR6808227_1.fastq | seqkit grep -s -i -p aggcg -P > seqkit_output/SRR6808227_1_sequences_w_AGGCG_positive_strand.fastq

]666;vte.shell.preexec!\


: 1

In [122]:
head -20 seqkit_output/SRR6808227_1_sequences_w_AGGCG_positive_strand.fastq

]666;vte.shell.preexec!\@SRR6808227.38 38 length=150
ATTACATTTAGGCGTAAAACTCATTTGTTTTTTGTTTGTTATAATGAAAGTCGTGTTGTTTAGTCAGTGAATTAGTATTAGACTTAAATCATATGGCTTGAATGAATTCGGCCTGATTTCGAAACCAAGCAGGTAGAGGCAGATATCGAA
+
AAFFFJAFJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJAJJJJJJJJJJJJJJJFJJJFJFJJJJJJJJJJJJJJFJJJJFJJJJJJJJJFJJJFFFFJJFFJJJJFJJJAFJ-7AFFJJJFJJJFJJF<<FAJJJJJJJ-A-A7FF
@SRR6808227.55 55 length=150
AAATCGAGCGCGGTCACGCTTATAGCTAGTAGAGGGACGCGCTCTAGCATCAACATCAGCCGCTACAGGCGCTTCCGGAGGAAGGAACGCCATCTTAGACCTTGAAGGGGAATATCCATATGTTGGAGGTTGATTATACGATGATGTACC
+
<AFFFA-<J7-7F<FJF-F7-7FJAJFJA-FAJJJ-7FJ<7FFJ<FJJJFF7FFJAJFFJJFF<A<JAAA-A<-7777F-77F-7F7J-7FA7F7-77AJJF-7-F-AAJFJ-FAFJ-7-77-AA--7<7<---77-AA7F-A--F-<FJ
@SRR6808227.65 65 length=150
CAAAGTGGCTGGAACATTCCCACCACGTAGGCGTGATTGGAGTGGTTATGTAGAGAAGGACACAGCTGGGCAGACAAATATATATTCAGTGGAGCCTGCGGTCTACGTGGCAGAGAGTGCACTCAGCTCAGGGACATCAGGGACATCAGC
+
AAFFFJFJJFJJJJJJJFJJJJJJJJJJJJJJJ7JJJFJJJJFAJJJJJJJJJJJJF<FJJJJJJJJJFFJJJFJJJJ<JFJJJJJJJJFJFJJJJAJJJFFA<JJFFAFJFFJJJFJFJJJJJJJ

: 1